xdsl/transforms/experimental/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,163 @@
+from collections.abc import Sequence
+
+from attr import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import memref, stencil, tensor
+from xdsl.dialects.builtin import IntegerAttr, IntegerType, ModuleOp, TensorType
+from xdsl.dialects.csl_ import csl_stencil
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, OpResult
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class ApplyOpAddPrefetch(RewritePattern):
+
+    prefetch_op: csl_stencil.PrefetchOp
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.ApplyOp, rewriter: PatternRewriter, /):
+
+        pass
+
+
+@dataclass(frozen=True)
+class AccessOpFromPrefetch(RewritePattern):
+
+    # index of the stencil access to be replaced by the prefetched buffer in the last arg
+    arg_index: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.AccessOp, rewriter: PatternRewriter, /):
+        assert len(op.offset) == 2
+        if tuple(op.offset) == (0, 0):
+            return
+        prefetched_arg = op.get_apply().region.block.args[-1]
```

Comment:
How do we know that the prefetched arg is always the last one?

Reply:
This is run by the `nested_rewriter` inside the `SwapToPrefetch` rewrite pattern. I've added comments to clarify.